test/common_distributed.py

```diff
@@ -129,26 +129,39 @@ def setUp(self):
         super(MultiProcessTestCase, self).setUp()
         self.skip_return_code_checks = []
         self.rank = self.MAIN_PROCESS_RANK
-        self.file = tempfile.NamedTemporaryFile(delete=False)
+        # spawn can not pass file handle, so pass file name here
+        self.file = tempfile.NamedTemporaryFile(delete=False).name
         self.processes = [self._spawn_process(rank) for rank in range(int(self.world_size))]
 
     def tearDown(self):
         super(MultiProcessTestCase, self).tearDown()
         for p in self.processes:
             p.terminate()
 
+    def _current_test_name(self):
+        # self.id() == e.g. '__main__.TestDistributed.TestAdditive.test_get_rank'
+        return self.id().split(".")[-1]
+
     def _spawn_process(self, rank):
         name = 'process ' + str(rank)
-        process = multiprocessing.Process(target=self._run, name=name, args=(rank,))
+        test_name = self._current_test_name()
+        process = multiprocessing.Process(
+            target=self.__class__._run,
+            name=name,
+            args=(test_name, rank, self.file)
+        )
         process.start()
         return process
 
-    def _run(self, rank):
+    @classmethod
+    def _run(cls, test_name, rank, file):
+        self = cls(test_name)
+
         self.rank = rank
+        self.file = file
 
-        # self.id() == e.g. '__main__.TestDistributed.test_get_rank'
-        # We're retreiving a corresponding test and executing it.
-        getattr(self, self.id().split(".")[2])()
+        # We're retrieving a corresponding test and executing it.
+        getattr(self, test_name)()
         sys.exit(0)
```

Comment:
Why exit(zero) explicitly shouldn't this return what is returned from the test ?

Reply:
@satgera I do not know the historic reason, but I guess put it here explicitly is no harm? if sys exit with other codes earlier, sys.exit(0) will not be executed; if sys.exit(0) is executed here, then it is fine as well

Reply:
@satgera 

I guess it's because the sub-process was forked, so they would run `self.tearDown()`, because the TestCase object was hold in a "TestRunner" object.

and sys.exit() was for skipping `self.tearDown()`.

However, with this PR, the TestCase object is not loaded in a test runner, `self.tearDown()` will not be run.

So I think it's safe to remove `sys.exit(0)` now.